### PR TITLE
Soloninko Andrey. Lab 2, opt. 1

### PR DESCRIPTION
**Pull request by Manymoera**
Inserts `instrument_start()` and `instrument_end()` functions into the begining and end of every function in a module.
Example: 
`void func() {}`
will be turned into
`void func() { instrument_start(); instrument_end(); }`

### CODE REVIEW

File: llvm/labs/lab2/soloninko_andrey/instrumentation.cpp [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+        (*module).getOrInsertFunction("instrument_start", type);
+    llvm::FunctionCallee func_l =
+        (*module).getOrInsertFunction("instrument_end", type);
+
+    bool foundInstrument_start = false;
+    bool foundInstrument_end = false;
+
+    for (llvm::BasicBlock &BB : F) {
+      for (llvm::Instruction &I : BB) {
+        if (auto *callInst = llvm::dyn_cast<llvm::CallInst>(&I)) {
```
Comment (m-ly4) [inline llvm/labs/lab2/soloninko_andrey/instrumentation.cpp]: In generic case it is more performant to just check that `F` is not present among users of `func_f` and `func_l` functions.

**Review by m-ly4 — CHANGES_REQUESTED** [review]

**Comment by Manymoera** [comment]
@m-ly4 I have changed my method of finding function calls. Now it searches for `F` to be present among users of `func_l` and `func_f` functions.

File: llvm/test/lab2/soloninko_andrey/foo.ll [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+  %0 = load i32, ptr %a.addr, align 4
+  %1 = load i32, ptr %b.addr, align 4
+  %add = add nsw i32 %0, %1
+  ret i32 %add
+}
+
+; CHECK-LABEL: @_Z6func_fv
+; CHECK: call void @instrument_start()
+; CHECL-NEXT: call void @instrument_end()
+; CHECK: call void @instrument_end()
```
Comment (m-ly4) [inline llvm/test/lab2/soloninko_andrey/foo.ll]:
```suggestion
; CHECK-NEXT: call void @instrument_end()
```

File: llvm/test/lab2/soloninko_andrey/foo.ll [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+define dso_local void @_Z6func_fv() #1 {
+entry:
+  call void @instrument_start()
+  call void @instrument_end()
+  ret void
+}
+
+; CHECK-LABEL: @_Z3foo_ifii
+; CHECK: call void @instrument_start()
+; CHECL-NEXT: %a.addr = alloca i32, align 4
```
Comment (m-ly4) [inline llvm/test/lab2/soloninko_andrey/foo.ll]:
```suggestion
; CHECK-NEXT: %a.addr = alloca i32, align 4
```

File: llvm/test/lab2/soloninko_andrey/foo.ll [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+; CHECK-NEXT: ret void
+define dso_local void @_Z5foo_vv() #1 {
+entry:
+  call void @instrument_start()
+  ret void
+}
+
+; CHECK-LABEL: @_Z8func_endii
+; CHECK: call void @instrument_start()
+; CHECL-NEXT: %a.addr = alloca i32, align 4
```
Comment (m-ly4) [inline llvm/test/lab2/soloninko_andrey/foo.ll]:
```suggestion
; CHECK-NEXT: %a.addr = alloca i32, align 4
```

File: llvm/test/lab2/soloninko_andrey/foo.ll [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+  %add = add nsw i32 %0, %1
+  store i32 %add, ptr %c, align 4
+  %2 = load i32, ptr %c, align 4
+  ret i32 %2
+}
+
+; CHECK-LABEL: @_Z5foo_vv
+; CHECK: call void @instrument_start()
+; CHECL-NEXT: call void @instrument_end()
+; CHECK: call void @instrument_end()
```
Comment (m-ly4) [inline llvm/test/lab2/soloninko_andrey/foo.ll]:
```suggestion
; CHECK-NEXT: call void @instrument_end()
```

File: llvm/test/lab2/soloninko_andrey/foo.ll [hunk 6]
```diff
@@ -0,0 +1,99 @@
+; RUN: opt -load-pass-plugin %llvmshlibdir/Instrumentation_Soloninko_Andrey_FIIT2%pluginext\
+; RUN: -passes=instrumentation -S %s | FileCheck %s
+
+; CHECK-LABEL: @_Z3fooii
+; CHECK: call void @instrument_start()
+; CHECL-NEXT: %a.addr = alloca i32, align 4
```
Comment (m-ly4) [inline llvm/test/lab2/soloninko_andrey/foo.ll]:
```suggestion
; CHECK-NEXT: %a.addr = alloca i32, align 4
```

**Review by m-ly4 — CHANGES_REQUESTED** [review]

**Review by m-ly4 — APPROVED** [review]

**Pull request merged**